src/buffer/out/Row.cpp

```diff
@@ -191,8 +347,331 @@ OutputCellIterator ROW::WriteCells(OutputCellIterator it, const til::CoordType i
+    // Safety: col cannot be incremented past _columnCount, because the last
+    // _charOffset at index _columnCount will never get the CharOffsetsTrailer flag.
+    while (_uncheckedIsTrailer(++col))
+    {
+    }
+    // Safety: col is now [1, _columnCount].
+    const auto end = _uncheckedCharOffset(col);
+
+    return { _chars.begin() + beg, _chars.begin() + end };
+}
+
+DbcsAttribute ROW::DbcsAttrAt(til::CoordType column) const noexcept
+{
+    const auto col = _clampedColumn(column);
+
+    auto attr = DbcsAttribute::Single;
+    // Safety: col is [0, _columnCount).
+    if (_uncheckedIsTrailer(col))
+    {
+        attr = DbcsAttribute::Trailing;
+    }
+    // Safety: col+1 is [1, _columnCount].
+    else if (_uncheckedIsTrailer(::base::strict_cast<size_t>(col) + 1u))
+    {
+        attr = DbcsAttribute::Leading;
+    }
+
+    return { attr };
+}
+
+std::wstring_view ROW::GetText() const noexcept
+{
+    return { _chars.data(), _charSize() };
+}
+
+DelimiterClass ROW::DelimiterClassAt(til::CoordType column, const std::wstring_view& wordDelimiters) const noexcept
+{
+    const auto col = _clampedColumn(column);
+    // Safety: col is [0, _columnCount).
+    const auto glyph = _uncheckedChar(_uncheckedCharOffset(col));
+
+    if (glyph <= L' ')
+    {
+        return DelimiterClass::ControlChar;
+    }
+    else if (wordDelimiters.find(glyph) != std::wstring_view::npos)
+    {
+        return DelimiterClass::DelimiterChar;
+    }
+    else
+    {
+        return DelimiterClass::RegularChar;
+    }
+}
+
+template<typename T>
+constexpr uint16_t ROW::_clampedUint16(T v) noexcept
+{
+    return static_cast<uint16_t>(std::max(T{ 0 }, std::min(T{ 65535 }, v)));
+}
+
+template<typename T>
+constexpr uint16_t ROW::_clampedColumn(T v) const noexcept
+{
+    return static_cast<uint16_t>(std::max(T{ 0 }, std::min<T>(_columnCount - 1u, v)));
+}
+
+template<typename T>
+constexpr uint16_t ROW::_clampedColumnInclusive(T v) const noexcept
+{
+    return static_cast<uint16_t>(std::max(T{ 0 }, std::min<T>(_columnCount, v)));
+}
+
+// Safety: col must be [0, _charSize()].
+wchar_t ROW::_uncheckedChar(size_t off) const noexcept
+{
+    return til::at(_chars, off);
+}
+
+uint16_t ROW::_charSize() const noexcept
+{
+    // Safety: _charOffsets is an array of `_columnCount + 1` entries.
+    return til::at(_charOffsets, _columnCount);
+}
+
+// Safety: col must be [0, _columnCount].
+uint16_t ROW::_uncheckedCharOffset(size_t col) const noexcept
+{
+    return til::at(_charOffsets, col) & CharOffsetsMask;
+}
+
+// Safety: col must be [0, _columnCount].
+bool ROW::_uncheckedIsTrailer(size_t col) const noexcept
+{
+    return WI_IsFlagSet(til::at(_charOffsets, col), CharOffsetsTrailer);
+}
+
+void ROW::_init() noexcept
+{
+    std::fill_n(_chars.begin(), _columnCount, UNICODE_SPACE);
+    iota_n(_charOffsets.begin(), _columnCount + 1u, uint16_t{ 0 });
+}
+
+void ROW::_resizeChars(uint16_t colExtEnd, uint16_t chExtBeg, uint16_t chExtEnd, size_t chExtEndNew)

```

Comment:
nit: It'd be nice if...
1. this was placed closer to `ReplaceCharacters()` since that's where this is used
2. there was a method description (or just more comments in general)